Big pull request 2

DavidEGrayson opened the pull request:
This pull request does a lot of stuff.  I think the most important things are:

- Detailed README
- More streamlined error handling with the `errors` package
- Support for updating fetchGitHub expressions
- Returning a non-zero exit code when there is an error.

expipiplus1 commented:
Looks amazing, I'll give it a read :)

expipiplus1 commented on README.md:
```diff
@@ -0,0 +1,59 @@
+        rev = "4f56fd184ef6020626492a6f954a486d54f8b7ba";
+        sha256 = "0nmyp5yrzl9dbq85wyiimsj9fklb8637a1936nw7zzvlnzkgh28n";
+      };
+    }
```
If one uses

<pre lang="no-highlight"><code>```nix
some {nix = expression;}
```</code></pre>


Then github does syntax highlighting :)

DavidEGrayson replied:
Ah, nice!

expipiplus1 commented on src/Update/Nix/FetchGit.hs:
```diff
@@ -23,66 +21,66 @@ import           Update.Span
 
--- | Gather the 'SpanUpdate's for a file at a particular filepath.
+-- | Given the contents of a Nix file, returns the SpanUpdates
+-- all the parts of the file we want to update.
```
The comment says that this function is getting the contents of the file, but the type says it's getting the path to the file.

expipiplus1 commented on src/Update/Nix/FetchGit.hs:
```diff
@@ -23,66 +21,66 @@ import           Update.Span
+  treeWithArgs <- hoistEither $ exprToFetchTree expr
+  treeWithLatest <- ExceptT $
+    sequenceA <$> mapConcurrently getFetchGitLatestInfo treeWithArgs
+  pure (fetchTreeToSpanUpdates treeWithLatest)
```
This is so much nicer than before!

expipiplus1 commented on src/Update/Nix/FetchGit.hs:
```diff
@@ -99,15 +97,12 @@ fetchTreeToSpanUpdates (FetchNode f) = [revUpdate, sha256Update]
-    days -> Just $ SpanUpdate (exprSpan versionExpr)
-                              ((quoteString . pack . show . maximum) days)
+maybeUpdateVersion node@(Node (Just versionExpr) _) = do
+  maxDay <- (maximumMay . fmap latestDate . universeBi) node
```
It would be nice to wrap the use of `universeBi` in an explicitly typed value, something like `allDatesInTree`.

DavidEGrayson replied:
What if I use `toList` (from Data.Foldable) instead of `universeBi`?  Would that be clear enough, since `toList` is not so flexible about what type it returns?  If not, I'm happy to add a helper function in Utils.hs:

``` haskell
treeToGitFetches :: FetchTree fetchInfo -> [fetchInfo]
treeToGitFetches = toList
```

expipiplus1 commented on src/Update/Nix/FetchGit/Utils.hs:
```diff
@@ -83,3 +102,21 @@ parseISO8601DateToDay t =
     Nothing -> Left $ InvalidDateString t
     Just day -> pure day
+
+instance Show Warning where
```
It's idiomatic to have GHC derive `Show`. The derived version has the nice property that it can be pasted into Haskell source to represent the shown value.

`show` as defined here could be renamed to `prettyShow`. The use of `print` earlier would have to be changed to `putStrLn . prettyShow`

DavidEGrayson replied:
I was looking in to this for a while.  That idiom is unavailable to us because our `Warning` type contains a `Text.PrettyPrint.ANSI.Leijen.Doc` object, which doesn't follow the idiom.  So when GHC derives `Show` for our `Warning`, it ends up using `show` on that doc object, which just returns a pretty-printed version suitable for console output.  So the output of `show` on a `Warning` looks something like this (which I just made up on the spot):

```
CouldNotParseInput foo.nix:14:5: got ',', expected '<' or '>'
  or '}' or '{'
```

The places that [talk about that idiom](https://hackage.haskell.org/package/base-4.9.0.0/docs/Text-Show.html) only refer to _derived_ instances of show, not manually-defined instances.

Still, I don't really care, so I'm happy to make a `prettyShow` function.  I think I would want to remove the `Show` typeclass from `Warning` so that we don't accidentally use it somewhere when trying to print a warning.

expipiplus1 replied:
yes, the prettyprint `Doc` object is ruining that :(

It's probably best not to have any instance of `Show` on `Warning` because of this.

expipiplus1 commented:
Does the fetching from Github work. I don't think that `nix-prefetch-git` is compatible, as `fetchFromGitHub` actually gets a `.zip` file.

DavidEGrayson commented:
Fetching from GitHub works.  fetchFromGitHub actually downloads a .tar.gz, but that's just an internal detail.  At the end of the day, fetchgit and fetchFromGitHub give you the same set of directories/files/symlinks.  Here is a shell session showing how fetchFromGitHub and nix-prefecth-git both give the same checksum of `19b0zyfsf0ja8qpv30vzi4kmj7w4n5v5ywafhm2idgc9zwlpyb74` for a particular repository:

```
$ cat default.nix 
let
  pkgs = import <nixpkgs> { };
in
  pkgs.fetchFromGitHub {
    owner = "b";
    repo = "hex";
    rev = "967f059ea49e45b7d0d82e2e1177ef860f5ec4bd";
    sha256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
  }

$ nix-build
these derivations will be built:
  /nix/store/r4ya9qbv44ynar3kfdrmr6zwbi5ph93s-hex-967f059ea49e45b7d0d82e2e1177ef860f5ec4bd-src.drv
building path(s) ‘/nix/store/dlk72zsin6l8hnlm5yw3d50zjcs10cig-hex-967f059ea49e45b7d0d82e2e1177ef860f5ec4bd-src’

trying https://github.com/b/hex/archive/967f059ea49e45b7d0d82e2e1177ef860f5ec4bd.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   147    0   147    0     0    435      0 --:--:-- --:--:-- --:--:--   463
100  140k  100  140k    0     0   107k      0  0:00:01  0:00:01 --:--:--  392k
unpacking source archive /tmp/nix-build-hex-967f059ea49e45b7d0d82e2e1177ef860f5ec4bd-src.drv-0/967f059ea49e45b7d0d82e2e1177ef860f5ec4bd.tar.gz
output path ‘/nix/store/dlk72zsin6l8hnlm5yw3d50zjcs10cig-hex-967f059ea49e45b7d0d82e2e1177ef860f5ec4bd-src’ has r:sha256 hash ‘19b0zyfsf0ja8qpv30vzi4kmj7w4n5v5ywafhm2idgc9zwlpyb74’ when ‘0mdqa9w1p6cmli6976v4wi0sw9r4p5prkj7lzfd1877wk11c9c73’ was expected
error: build of ‘/nix/store/r4ya9qbv44ynar3kfdrmr6zwbi5ph93s-hex-967f059ea49e45b7d0d82e2e1177ef860f5ec4bd-src.drv’ failed

$ nix-prefetch-git https://github.com/b/hex 967f059ea49e45b7d0d82e2e1177ef860f5ec4bd
Initialized empty Git repository in /tmp/git-checkout-tmp-aJgppNIA/hex-967f059/.git/
remote: Counting objects: 11, done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 11 (delta 0), reused 11 (delta 0), pack-reused 0
Unpacking objects: 100% (11/11), done.
From https://github.com/b/hex
 * branch            HEAD       -> FETCH_HEAD
Switched to a new branch 'fetchgit'
removing `.git'...

git revision is 967f059ea49e45b7d0d82e2e1177ef860f5ec4bd
path is /nix/store/6g619xqcdq13yzl7p2zcf5aplxls7q1c-hex-967f059
git human-readable version is -- none --
Commit date is 2012-10-23 11:12:48 -0700
hash is 19b0zyfsf0ja8qpv30vzi4kmj7w4n5v5ywafhm2idgc9zwlpyb74
{
  "url": "https://github.com/b/hex",
  "rev": "967f059ea49e45b7d0d82e2e1177ef860f5ec4bd",
  "date": "2012-10-23T11:12:48-07:00",
  "sha256": "19b0zyfsf0ja8qpv30vzi4kmj7w4n5v5ywafhm2idgc9zwlpyb74"
}
```

DavidEGrayson commented:
I addressed all your comments fully except the implementation of `maybeUpdateVersion`.  If you're happy with what I did in `maybeUpdateVersion` (changing `universeBi` to `toList`) then this is good to be merged!

expipiplus1 commented:
Great! merging now :)

expipiplus1 commented:
Thank you!
